lib/services/graph-progress.js

```diff
@@ -0,0 +1,230 @@
+// Copyright © 2017 Dell Inc. or its subsidiaries.  All Rights Reserved.
+
+'use strict';
+
+module.exports = graphProgressServiceFactory;
+
+graphProgressServiceFactory.$provide = 'Services.GraphProgress';
+graphProgressServiceFactory.$inject = [
+    '_',
+    'Protocol.Events',
+    'Services.Waterline',
+    'Logger',
+    'Constants',
+    'Promise',
+    'Assert',
+    'GraphProgress'
+];
+
+function graphProgressServiceFactory(
+    _,
+    eventsProtocol,
+    waterline,
+    Logger,
+    Constants,
+    Promise,
+    assert,
+    GraphProgress
+) {
+    var logger = Logger.initialize(graphProgressServiceFactory);
+
+    function GraphProgressService() {
+    }
+
+    /**
+     * Publish a graph started progress event over AMQP.
+     *
+     * @param {Object} graph
+     * @returns {Promise}
+     * @memberOf GraphProgressService
+     */
+    GraphProgressService.prototype.publishGraphStarted = function(graph) {
+        assert.object(graph, 'graph');
+        assert.string(graph.name , 'graph.name');
+
+        var progress = GraphProgress.create(graph, 'Graph "' + graph.name + '" started');
+        return eventsProtocol.publishProgressEvent(
+            graph.instanceId,
+            progress.getProgressEventData()
+        )
+        .catch(function(error) {
+            logger.error('Error publishing graph started progress event', {
+                graphId: graph.instanceId,
+                _status: graph._status,
+                error: error
+            });
+        });
+    };
+
+    /**
+     * Publish a graph finished progress event over AMQP.
+     *
+     * @param {Object} graph
+     * @param {String} graphState
+     * @returns {Promise}
+     * @memberOf GraphProgressService
+     */
+    GraphProgressService.prototype.publishGraphFinished = function(graph, graphState) {
```

Comment:
In fact, we have use case that the graph is ongoing, but need to publish the finished event ahead of time. So we just leave the validation of graph status to caller of the function.